File: android_fragment/commons-core/src/main/java/com/common/core/utils/DateUtils.java
package com.common.core.utils;

import android.annotation.SuppressLint;
import android.os.SystemClock;

import java.net.URL;
import java.net.URLConnection;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * 时间工具类
 */
@SuppressLint("SimpleDateFormat")
public class DateUtils {
    private static SimpleDateFormat simpleFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
    private static SimpleDateFormat simpleFormat2 = new SimpleDateFormat("yyyy-MM-dd HH:mm");
    private static SimpleDateFormat MMddSimpleFormat = new SimpleDateFormat("MM月dd日");
    private static long internetTime;
    private static long uptimeMillis;

    /**
     * 时间保持和服务器时间同步
     *
     * @param dateStr
     */
    public static void setHeaderDate(String dateStr) {
        SimpleDateFormat greenwichDate = new SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss 'GMT'", Locale.US);
        Calendar cal = Calendar.getInstance();
        // 时区设为格林尼治
        greenwichDate.setTimeZone(TimeZone.getTimeZone("GMT"));
        try {
            Date date = greenwichDate.parse(dateStr);
            setInternetTime(date.getTime());
            CmLog.e("DateUtils", formatDate(date));
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    public static void setInternetTime(long internetTime) {
        DateUtils.internetTime = internetTime;
        DateUtils.uptimeMillis = SystemClock.uptimeMillis();
    }

    /**
     * 取网络时间
     *
     * @return
     */
    public static long getInternetTime() {
        if (internetTime <= 0) {
            long time = System.currentTimeMillis();
            try {
                URL url = new URL("http://www.baidu.com");
                URLConnection uc = url.openConnection();
                uc.connect();
                time = uc.getDate();
            } catch (Exception e) {
                e.printStackTrace();
            }
            return time;
        } else {
            return internetTime + (SystemClock.uptimeMillis() - uptimeMillis);
        }
    }

    /**
     * 获取网络时间
     *
     * @return
     */
    public static Date getNetWorkDate() {
        return getNetWorkDate(false);
    }

    /**
     * @param isMust 是否必须获取
     * @return 事件
     * @Description: 获取网络时间
     */
    public static Date getNetWorkDate(boolean isMust) {
        Date date = new Date(getInternetTime());
        return date;
    }

    /**
     * 获取网络事件
     *
     * @return
     */
    public static Calendar getNetWorkCalendar() {
        Date date = DateUtils.getNetWorkDate();
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        return calendar;
    }

    public static String formatYMD(long timestamp) {
        long time = timestamp * 1000;
        Date date = new Date(time);
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
        return format.format(date);
    }

    public static String formatTime(long millisecond) {
        Date date = new Date(millisecond);
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
        return format.format(date);
    }

    public static String formatMD(Date date) {
        return MMddSimpleFormat.format(date);
    }

    public static String formatHHMM(long timestamp) {
        long time = timestamp * 1000;
        Date date = new Date(time);
        SimpleDateFormat format = new SimpleDateFormat("HH:mm");
        return format.format(date);
    }

    public static String formatDate(Date date) {
        return simpleFormat.format(date);
    }

    public static Date parseDate(String dateStr) {
        try {
            return simpleFormat.parse(dateStr);
        } catch (ParseException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Date parseDate2(String dateStr) {
        try {
            return simpleFormat2.parse(dateStr);
        } catch (ParseException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static String formatDate(long timemillis) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(timemillis);
        return simpleFormat.format(calendar.getTime());
    }

    public static String formatNow() {
        return formatDate(new Date());
    }

    public static String getYYYYMM(Calendar calendar) {
        SimpleDateFormat simpleFormat = new SimpleDateFormat("yyyyMM");
        return simpleFormat.format(calendar.getTime());
    }

    public static String getYyyyMM(Date date) {
        SimpleDateFormat yyyyMMSimpleFormat = new SimpleDateFormat("yyyy月MM月");
        return yyyyMMSimpleFormat.format(date);
    }

    /**
     * 返回时间
     *
     * @param datetime
     */
    public static Calendar getCalendar(String datetime) {
        try {
            Date date = simpleFormat.parse(datetime);
            Calendar calendar = Calendar.getInstance();
            calendar.setTime(date);
            return calendar;
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * 获取天数
     *
     * @param year  年份
     * @param month 月份
     * @return
     */
    public static int getDayNumber(int year, int month) {
        int dayNumber;
        switch (month + 1) {
            case 4:
            case 6:
            case 9:
            case 11:
                dayNumber = 30;
                break;
            case 2:
                if ((year % 100 == 0 && year % 400 == 0) || year % 100 != 0 && year % 4 == 0) {
                    dayNumber = 29;
                } else {
                    dayNumber = 28;
                }
                break;
            default:
                dayNumber = 31;
                break;
        }
        return dayNumber;
    }

    public static int getYearNumber(int year) {
        if ((year % 100 == 0 && year % 400 == 0) || year % 100 != 0 && year % 4 == 0) {
            return 366;
        } else {
            return 365;
        }
    }

    /**
     * 时间大于时间2多少天
     *
     * @param year
     * @param month
     * @param day
     * @param year2
     * @param month2
     * @param day2
     * @return
     */
    public static int after(int year, int month, int day, int year2, int month2, int day2) {
        if (year < year2) return -1;
        if (year == year2) {
            if (month < month2) return -1;
            if (month == month2) {
                if (day < day2) return -1;
                return day - day2;
            } else {
                int number = getDayNumber(year2, month2) - day2;
                for (int i = month2 + 1; i < month; i++) {
                    number += getDayNumber(year, i);
                }
                number += day;
                return number;
            }
        } else {
            int number = getDayNumber(year2, month2) - day2;
            for (int m = month2 + 1; m < 12; m++) {
                number += getDayNumber(year2, m);
            }
            for (int y = year2 + 1; y < year; y++) {
                number += getYearNumber(y);
            }
            for (int m = 0; m < month; m++) {
                number += getDayNumber(year, m);
            }
            number += day;
            return number;
        }
    }

    public static boolean afterDay(Calendar calendar, Calendar calendar2) {
        int year = calendar.get(Calendar.YEAR);
        int day = calendar.get(Calendar.DAY_OF_YEAR);
        int year2 = calendar2.get(Calendar.YEAR);
        int day2 = calendar2.get(Calendar.DAY_OF_YEAR);
        if (year > year2 || (year == year2 && day > day2)) return true;
        return false;
    }
}

File: example/commons-core/src/main/java/com/common/core/utils/DateUtils.kt
package com.common.core.utils

import android.annotation.SuppressLint
import android.os.SystemClock
import com.common.core.utils.DateUtils.getInternetTime

import java.net.URL
import java.net.URLConnection
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * 时间工具类
 */
@SuppressLint("SimpleDateFormat")
object DateUtils {
    private val simpleFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    private val simpleFormat2 = SimpleDateFormat("yyyy-MM-dd HH:mm")
    private val MMddSimpleFormat = SimpleDateFormat("MM月dd日")
    private var internetTime: Long = 0
    private var uptimeMillis: Long = 0

    /**
     * 获取网络时间
     *
     * @return
     */
    val netWorkDate: Date
        get() = getNetWorkDate(false)

    /**
     * 获取网络事件
     *
     * @return
     */
    val netWorkCalendar: Calendar
        get() {
            val date = DateUtils.netWorkDate
            val calendar = Calendar.getInstance()
            calendar.time = date
            return calendar
        }

    /**
     * 时间保持和服务器时间同步
     *
     * @param dateStr
     */
    fun setHeaderDate(dateStr: String) {
        val greenwichDate = SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss 'GMT'", Locale.US)
        val cal = Calendar.getInstance()
        // 时区设为格林尼治
        greenwichDate.timeZone = TimeZone.getTimeZone("GMT")
        try {
            val date = greenwichDate.parse(dateStr)
            setInternetTime(date.time)
            CmLog.e("DateUtils", formatDate(date))
        } catch (e: ParseException) {
            e.printStackTrace()
        }

    }

    fun setInternetTime(internetTime: Long) {
        DateUtils.internetTime = internetTime
        DateUtils.uptimeMillis = SystemClock.uptimeMillis()
    }

    /**
     * 取网络时间
     *
     * @return
     */
    fun getInternetTime(): Long {
        if (internetTime <= 0) {
            var time = System.currentTimeMillis()
            try {
                val url = URL("http://www.baidu.com")
                val uc = url.openConnection()
                uc.connect()
                time = uc.date
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return time
        } else {
            return internetTime + (SystemClock.uptimeMillis() - uptimeMillis)
        }
    }

    /**
     * @param isMust 是否必须获取
     * @return 事件
     * @Description: 获取网络时间
     */
    fun getNetWorkDate(isMust: Boolean): Date {
        return Date(getInternetTime())
    }

    fun formatYMD(timestamp: Long): String {
        val time = timestamp * 1000
        val date = Date(time)
        val format = SimpleDateFormat("yyyy-MM-dd")
        return format.format(date)
    }

    fun formatTime(millisecond: Long): String {
        val date = Date(millisecond)
        val format = SimpleDateFormat("yyyy-MM-dd")
        return format.format(date)
    }

    fun formatMD(date: Date): String {
        return MMddSimpleFormat.format(date)
    }

    fun formatHHMM(timestamp: Long): String {
        val time = timestamp * 1000
        val date = Date(time)
        val format = SimpleDateFormat("HH:mm")
        return format.format(date)
    }

    fun formatDate(date: Date): String {
        return simpleFormat.format(date)
    }

    fun parseDate(dateStr: String): Date? {
        try {
            return simpleFormat.parse(dateStr)
        } catch (e: ParseException) {
            e.printStackTrace()
            return null
        }

    }

    fun parseDate2(dateStr: String): Date? {
        try {
            return simpleFormat2.parse(dateStr)
        } catch (e: ParseException) {
            e.printStackTrace()
            return null
        }

    }

    fun formatDate(timemillis: Long): String {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = timemillis
        return simpleFormat.format(calendar.time)
    }

    fun formatNow(): String {
        return formatDate(Date())
    }

    fun getYYYYMM(calendar: Calendar): String {
        val simpleFormat = SimpleDateFormat("yyyyMM")
        return simpleFormat.format(calendar.time)
    }

    fun getYyyyMM(date: Date): String {
        val yyyyMMSimpleFormat = SimpleDateFormat("yyyy月MM月")
        return yyyyMMSimpleFormat.format(date)
    }

    /**
     * 返回时间
     *
     * @param datetime
     */
    fun getCalendar(datetime: String): Calendar? {
        try {
            val date = simpleFormat.parse(datetime)
            val calendar = Calendar.getInstance()
            calendar.time = date
            return calendar
        } catch (e: Exception) {
            return null
        }

    }

    /**
     * 获取天数
     *
     * @param year  年份
     * @param month 月份
     * @return
     */
    fun getDayNumber(year: Int, month: Int): Int {
        val dayNumber: Int
        when (month + 1) {
            4, 6, 9, 11 -> dayNumber = 30
            2 -> if (year % 100 == 0 && year % 400 == 0 || year % 100 != 0 && year % 4 == 0) {
                dayNumber = 29
            } else {
                dayNumber = 28
            }
            else -> dayNumber = 31
        }
        return dayNumber
    }

    fun getYearNumber(year: Int): Int {
        return if (year % 100 == 0 && year % 400 == 0 || year % 100 != 0 && year % 4 == 0) {
            366
        } else {
            365
        }
    }

    /**
     * 时间大于时间2多少天
     *
     * @param year
     * @param month
     * @param day
     * @param year2
     * @param month2
     * @param day2
     * @return
     */
    fun after(year: Int, month: Int, day: Int, year2: Int, month2: Int, day2: Int): Int {
        if (year < year2) return -1
        if (year == year2) {
            if (month < month2) return -1
            if (month == month2) {
                return if (day < day2) -1 else day - day2
            } else {
                var number = getDayNumber(year2, month2) - day2
                for (i in month2 + 1 until month) {
                    number += getDayNumber(year, i)
                }
                number += day
                return number
            }
        } else {
            var number = getDayNumber(year2, month2) - day2
            for (m in month2 + 1..11) {
                number += getDayNumber(year2, m)
            }
            for (y in year2 + 1 until year) {
                number += getYearNumber(y)
            }
            for (m in 0 until month) {
                number += getDayNumber(year, m)
            }
            number += day
            return number
        }
    }

    fun afterDay(calendar: Calendar, calendar2: Calendar): Boolean {
        val year = calendar.get(Calendar.YEAR)
        val day = calendar.get(Calendar.DAY_OF_YEAR)
        val year2 = calendar2.get(Calendar.YEAR)
        val day2 = calendar2.get(Calendar.DAY_OF_YEAR)
        return if (year > year2 || year == year2 && day > day2) true else false
    }
}
